app/controllers/studies_controller.rb

```diff
@@ -180,92 +123,24 @@ def sync_study
-    # them out at the end as removing them from the list mid-iteration will skip the next file accidentally
-    orphaned_directories = []
-    @directories.each do |directory|
-      synced = true
-      if @files_by_dir[directory.name].present?
-        remove_from_dir = []
-        directory.files.each do |file|
-          if @files_by_dir[directory.name].detect {|f| f['generation'].to_i == file['generation'].to_i}.nil?
-            synced = false
-            remove_from_dir << file
-          else
-            next
-          end
-        end
-        # if no longer synced, check if already in the list and remove as files list has changed
-        if !synced
-          directory.sync_status = false
-          # remove files that are no longer present
-          remove_from_dir.each do |remove|
-            directory.files.delete(remove)
-          end
-          @unsynced_directories.delete_if {|dir| dir.name == directory.name}
-          if directory.files.any?
-            @unsynced_directories << directory
-          else
-            orphaned_directories << directory # queue for deletion as this directory is empty now
-          end
-        elsif directory.sync_status
-          @synced_directories << directory
-        end
-      else
-        # directory did not exist in @files_by_dir, so this directory_listing is orphaned and should be removed
-        orphaned_directories << directory
-      end
-    end
-
-    # remove orphaned directories
-    orphaned_directories.each do |orphan|
-      orphan.destroy
-    end
-
-    # provisionally save unsynced directories so we don't have to pass huge arrays of filenames/sizes in the form
-    # users clicking "don't sync" actually delete entries
-    @unsynced_directories.each do |directory|
-      directory.save
-    end
+    # refresh study state before continuing as new records may have been inserted
+    @study.reload
 
-    # reload unsynced directories to remove any that were orphaned
+    @synced_directories = @study.directory_listings.are_synced
     @unsynced_directories = @study.directory_listings.unsynced
 
     # split directories into primary data types and 'others'
     @unsynced_primary_data_dirs = @unsynced_directories.select {|dir| DirectoryListing::PRIMARY_DATA_TYPES.include?(dir.file_type)}
     @unsynced_other_dirs = @unsynced_directories.select {|dir| !DirectoryListing::PRIMARY_DATA_TYPES.include?(dir.file_type)}
 
     # now determine if we have study_files that have been 'orphaned' (cannot find a corresponding bucket file)
-    @orphaned_study_files = @study_files - @synced_study_files
-    @available_files = @unsynced_files.map {|f| {name: f.name, generation: f.generation, size: f.upload_file_size}}
+    @orphaned_study_files = StudySyncService.find_orphaned_files(@study)
+    @available_files = @unsynced_files.map { |f| { name: f.name, generation: f.generation, size: f.upload_file_size } }
+    @synced_study_files = @study.study_files.valid - @orphaned_study_files
```

Comment:
The reason for the duplication is because the controllers need to set those variables locally to be available for the response.  It's a somewhat similar reason to the above comment about the error tracking duplication.  I'll take a crack at DRYing this up, but I'm not sure we'd end up with anything that's better or more readable.

Reply:
Actually there is some small refinement that can be done here to reduce duplication.  I'll re-request review when it's done.

Reply:
[dc8c94b](https://github.com/broadinstitute/single_cell_portal_core/pull/1598/commits/dc8c94bef37510eec18c7a179227e4a1e2a9fb5d)